### PR TITLE
KAFKA-8319: Make KafkaStreamsTest a non-integration test class

### PR DESCRIPTION
**Pull request by guozhangwang**
Previous KafkaStreamsTest takes 2min20s on my local laptop, because lots of its integration test which is producing / consuming records, and checking state directory file system takes lots of time. On the other hand, these tests should be well simplified with mocks.

This test reduces the test from a clumsy integration test class into a unit tests with mocks of its internal modules. And some other test functions should not be in KafkaStreamsTest actually and have been moved to other modular test classes. Now it takes 2s.

Also it helps removing the potential flakiness of the following (some of them are claimed resolved only because we have not seen them recently, but after looking at the test code I can verify they are still flaky):

* KAFKA-5818 (the original JIRA ticket indeed exposed a real issue that has been fixed, but the test itself remains flaky)
* KAFKA-6215
* KAFKA-7921
* KAFKA-7990
* KAFKA-8319
* KAFKA-8427

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:463]: This is a piggy-backed fix of MockConsumer: in the actual KafkaConsumer we do allow the `close()` call to be triggered multiple times, so there's no point checking this.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Similar to stream threads, making it accessible directly from KafkaStreamsTest.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:120]: These are moved to StreamsConfigTest.

File: streams/src/test/java/org/apache/kafka/test/MockClientSupplier.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/test/MockClientSupplier.java:61]: This is a piggy-backed fix: the controller's id is defined as -1, so with a single node cluster we should set it's controller as -1 not as 0.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:226]: See the other comment below in `MockClientSupplier.java`

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:586]: This is moved to StreamThreadTest.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:632]: These are examples where we should really not rely on checking the underlying state directory files, but just check the mocked state-directory as the expected construction call, etc.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:774]: See the other comment above: instead of checking the file system which is costly and flaky, we should just use a mock with expected constructors / function calls.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:832]: Moved up in the file, no logical changes.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:455]: This is a flaky test that should be removed: the streams.start() would only throw if we are still in the REBALANCING state, which is not really a good test anyways.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:375]: These two tests are not really for KafkaStreams but for StreamThreads / GlobalStreamThread.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: package-private should also work, right?

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: What's up with this last one? Should it be in the finally block?

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:375]: Are these cases already covered by existing tests over there? I didn't see them moved as part of this PR.

**Review by vvcephei — COMMENTED** [review]
Hey @guozhangwang ,

Thanks so much for this. I made a quick pass (definitely didn't go over all the tests to make sure they are effectively the same), and everything looks reasonable. I left a few comments.

Thanks,
-John

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Yes!

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:405]: This comment is incorrect too :) it should be `make sure we do NOT have global state thread`.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:375]: The first test is covered in `GlobalStreamThreadTest#shouldThrowStreamsExceptionOnStartupIfExceptionOccurred`: we call partitionsFor during the initialization of the global task, which would throw if the broker is not connected.

The second test is covered in `shouldNotThrowWhenPendingShutdownInRunOnce` and `shouldNotThrowWithoutPendingShutdownInRunOnce` of `StreamThreadTest` to me.

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:375]: sounds good! Thanks for the confirmation.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: If it is no more an integration test, this should be removed.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:86]: Why do you need to prepare `KafkaStreams` for testing? Only classes that are mocked and that are either `final` or have static methods need to be prepared.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Not sure I follow the logic of this test... We want to show that it doesn't block, but then we just assert that the call doesn't take longer than 30 seconds.

It seems like this test is sensitive to timing, which is not ideal. I guess that what we really wanted to test is that the close method never calls `Thread.sleep()`/`Thread.join()` or maybe just never calls a specific list of blocking methods on the clients. Now that we have all the mocks above, it seems like we can make this assertion directly.

**Review by vvcephei — APPROVED** [review]
Hey @guozhangwang ,

I made another, closer, pass. Everything looks good to me. I have a concern about one test, but my concern pre-dates this PR. I'll leave it to you whether you want to take the feedback or not. Regardless, your PR should already be a huge help.

Thanks,
-John

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Do we still want this labeled as an integration test?

**Review by bbejeck — APPROVED** [review]
Thanks for this PR @guozhangwang, this will really help with the length of time tests take.  I just have one minor comment otherwise this LGTM.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: I agree, I've actually simplify the logic to just check that `close()` call returns false, since with mock-time it would never elapse any time, and if the call ever waits on the state transition it would not return ever.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: ack.

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: ack.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:86]: We actually need to prepare `KafkaStreams` for cases like metrics reporter registration.

**Pull request merged**
